src/compiler/utilities.ts

```diff
@@ -1904,6 +1904,80 @@ namespace ts {
         });
     }
 
+    /**
+     * Detached comment is a comment at the top of file or function body that is separated from
+     * the next statement by space.
+     */
+    export function emitDetachedComments(currentSourceFile: SourceFile, writer: EmitTextWriter,
+        writeComment: (currentSourceFile: SourceFile, writer: EmitTextWriter, comment: CommentRange, newLine: string) => void,
+        node: TextRange, newLine: string, removeComments: boolean,
+        detachedCommentsInfo: { nodePos: number; detachedCommentEndPos: number }[]) {

```

Comment:
I think it will be better to not pass in the `detachedCommentsInfo` array into the function and return `currentDetachedCommentInfo ` and then have a function in emitter call `emitDetachedCommentsAndUpdateCommentsInfo` that calls into this function and do the following logic 
```ts
if (detachedCommentsInfo) {
   detachedCommentsInfo.push(currentDetachedCommentInfo);
}
else {
  detachedCommentsInfo = [currentDetachedCommentInfo];
}
```
So that in the case of using the function in declarationEmitter.ts, we will not be doing throw away work.